Add support for NSS hosts database lookup.

scabrero opened the pull request:
This PR adds support to cache and answer NSS hosts requests for the proxy, LDAP and AD providers, enumeration included.

As hosts and networks does not fit into the id_provider, it adds a new data provider target configurable by the new resolver_provider parameter.

Fixes https://pagure.io/SSSD/sssd/issue/359

sumit-bose commented:
Hi @scabrero,

thank you very much for this substantial contribution.

May I asked you to fix the Python pep8 issues identified by the CI test so that the other CI tests can be run?

For your convenience I add them here as well:

```
Start: Mon Oct 28 15:41:33 UTC 2019
+ find . -path ./src/config -prune -o -name '*.py' -exec pep8 --ignore=E121,E123,E126,E226,E24,E704,W503,W504,W605,E722 '{}' +
./src/tests/intg/ds_openldap.py:278:80: E501 line too long (84 > 79 characters)
./src/tests/intg/test_resolver.py:37:1: E265 block comment should start with '# '
./src/tests/intg/test_resolver.py:41:1: E302 expected 2 blank lines, found 1
./src/tests/intg/test_resolver.py:77:80: E501 line too long (84 > 79 characters)
./src/tests/intg/test_resolver.py:208:32: E251 unexpected spaces around keyword / parameter equals
./src/tests/intg/test_resolver.py:208:34: E251 unexpected spaces around keyword / parameter equals
./src/tests/intg/test_resolver.py:213:32: E251 unexpected spaces around keyword / parameter equals
./src/tests/intg/test_resolver.py:213:34: E251 unexpected spaces around keyword / parameter equals
./src/tests/intg/sssd_hosts.py:31:1: E302 expected 2 blank lines, found 1
./src/tests/intg/sssd_hosts.py:38:1: E302 expected 2 blank lines, found 1
./src/tests/intg/sssd_nss.py:32:1: E302 expected 2 blank lines, found 1
./src/tests/intg/sssd_nss.py:51:1: E302 expected 2 blank lines, found 1
./src/tests/intg/sssd_nss.py:60:1: E302 expected 2 blank lines, found 1
./src/tests/intg/sssd_nss.py:70:1: E302 expected 2 blank lines, found 1
./src/tests/intg/ldap_ent.py:150:26: E201 whitespace after '('
./src/tests/intg/ldap_ent.py:150:57: E202 whitespace before ')'
End: Mon Oct 28 15:41:36 UTC 2019
```

bye,
Sumit

scabrero commented:
Hi @sumit-bose, I fixed the PEP8 errors, the remaining CI failure seems unrelated to this PR.

sumit-bose commented:
Hi,

just a short update, I'm reviewing and testing the patch set, but I'm not done yet.

bye,
Sumit

scabrero commented:
@sumit-bose, any comments so far? May I rebase the branch due to the merge conflict or do you prefer to finish the review first? I would like to start with networks lookup if everything looks good to you.

sumit-bose commented on src/providers/ldap/ldap_opts.c:
```diff
@@ -368,6 +370,14 @@ struct sdap_attr_map service_map[] = {
+    { "ldap_iphost_name", "cn", SYSDB_NAME, NULL },
+    { "ldap_iphost_number", "ipHostNumber", SYSDB_IP_HOST_ATTR_ADDRESS, NULL },
+    { "ldap_iphost_entry_usn", NULL, SYSDB_USN, NULL },
+    SDAP_ATTR_MAP_TERMINATOR
```
It looks like man page entries for the new attribute options are missing.

sumit-bose commented on src/responder/nss/nss_cmd.c:
```diff
@@ -451,6 +451,61 @@ static errno_t nss_getby_sid(struct cli_ctx *cli_ctx,
+    }
+
+    subreq = nss_get_object_send(cmd_ctx, cli_ctx->ev, cli_ctx,
+                                 data, SSS_MC_NONE, NULL, 0);
```
I think it would be better to use the `memcache` parameter here instead of just `SSS_MC_NONE`.

sumit-bose commented on src/responder/common/responder_dp.c:
```diff
@@ -327,3 +327,128 @@ sss_dp_get_account_recv(TALLOC_CTX *mem_ctx,
+    }
+
+    /* either, or, not both */
+    if (name != NULL && address != 0) {
```
Since `address` is a char pointer as well please use `address != NULL`.

The case where both are NULL is not handled here, is this expected?

Since both parameter are of the same type, did you consider to use a single char pointer for the value and a bool or enum to indicate if it is a name or an address? I know that this style is used for other requests as well, but typically in those requests the parameters are of different types (char pointer and some integer type). You do not have to change this I'm just wondering if you tried to be similar to other request or if there might be even benefits doing it this way.

sumit-bose commented on src/util/util.c:
```diff
@@ -1073,3 +1073,47 @@ int sss_rand(void)
+        }
+
+        pcanon = inet_ntop(AF_INET6, binaddr, *canonical_address,
+                           INET6_ADDRSTRLEN);
```
Have you considered to use `getaddrinfo()`/`getnameinfo()` with flag `NI_NUMERICHOST` which should be able to transparently handle both address types?

scabrero commented:
@sumit-bose is there any way to trigger a new CI run? All tests are passing in my computer.

sumit-bose commented:
retest this, please

alexey-tikhonov commented:
> @sumit-bose is there any way to trigger a new CI run? All tests are passing in my computer.

Hi,

I restarted PR CI.

But having a quick glance at results of previous run, there are some compilation errors like:
```
/shared/sssd/src/tests/cmocka/common_mock_resp_dp.c:66:1: error: conflicting types for 'sss_dp_resolver_get_send'
   66 | sss_dp_resolver_get_send(TALLOC_CTX *mem_ctx,
      | ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /shared/sssd/src/tests/cmocka/common_mock_resp_dp.c:24:
/shared/sssd/src/responder/common/responder.h:300:1: note: previous declaration of 'sss_dp_resolver_get_send' was here
  300 | sss_dp_resolver_get_send(TALLOC_CTX *mem_ctx,
      | ^~~~~~~~~~~~~~~~~~~~~~~~
```

scabrero commented:
I didn't have cmoka installed and the file was not built. Everything is OK now.

pbrezina commented:
Thank you. I just gave it a quick look, especially the cache_req parts per Sumit's requests and all looks good to me. Though I haven't gone through all the patches thoroughly so I'll leave it for Sumit to ack it.

pbrezina commented:
@sumit-bose Bump. Can you re-review the patches please?

sumit-bose commented on src/providers/ldap/ldap_resolver_cleanup.c:
```diff
@@ -0,0 +1,152 @@
+    TALLOC_CTX *tmp_ctx;
+    struct sdap_id_ctx *id_ctx;
+    struct sdap_domain *sdom;
+    bool in_transaction;
```
Please initialize `in_transaction` as `false`

sumit-bose commented on src/providers/proxy/proxy_hosts.c:
```diff
@@ -21,14 +21,308 @@
+        }
+
+        /* Skip duplicates */
+        for (j = 0; j < len && addresses[j] != NULL; j++) {
```
There is a clang warning that `Array access (from variable 'addresses') results in a null pointer dereference` (same for `aliases` below as well). I think `j < len` should prevent this but maybe you can add a NULL check to make clang happy?

sumit-bose commented on src/providers/proxy/proxy_hosts.c:
```diff
@@ -21,14 +21,308 @@
+        DEBUG(SSSDBG_TRACE_INTERNAL, "Host [%s] not found, removing from "
+              "cache\n", name);
+        sysdb_host_delete(domain, search_name, NULL);
+        ret = ENOENT;
```
Is `goto done;` missing here?

sumit-bose commented on src/providers/proxy/proxy_hosts.c:
```diff
@@ -21,14 +21,308 @@
+        /* FALLTHROUGH */
+
+    default:
+        ret = ENOTSUP;
```
`ret` is not used in the `default` branch.

sumit-bose commented on src/sss_client/nss_hosts.c:
```diff
@@ -0,0 +1,442 @@
+    data_len = sizeof(uint32_t) + sizeof(socklen_t) + addrlen;
+    data = malloc(data_len);
+    if (data == NULL) {
+        ret = ENOMEM;
```
`ret` is not used in the branch.

sumit-bose commented:
Hi,

thank you for your patience.

The inline comments are Coverity and Clang finding.

Additionally please remve the vim tags

```
$ git grep 'vim:' ../
../src/db/sysdb_iphosts.c:/* vim: set expandtab ts=4 sw=4: */
../src/db/sysdb_iphosts.h:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ad/ad_resolver.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ad/ad_resolver.h:/* vim: set expandtab ts=4 sw=4: */
../src/providers/data_provider/dp_target_resolver.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/ldap_resolver_cleanup.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/ldap_resolver_enum.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/ldap_resolver_enum.h:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/sdap_async_iphost.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/sdap_async_resolver_enum.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/sdap_async_resolver_enum.h:/* vim: set expandtab ts=4 sw=4: */
../src/providers/ldap/sdap_iphost.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/proxy/proxy_hosts.c:/* vim: set expandtab ts=4 sw=4: */
../src/providers/proxy/proxy_init.c:/* vim: set expandtab ts=4 sw=4: */
../src/responder/common/cache_req/plugins/cache_req_enum_ip_hosts.c:/* vim: set expandtab ts=4 sw=4: */
../src/responder/common/cache_req/plugins/cache_req_ip_host_by_addr.c:/* vim: set expandtab ts=4 sw=4: */
../src/responder/common/cache_req/plugins/cache_req_ip_host_by_name.c:/* vim: set expandtab ts=4 sw=4: */
../src/sss_client/nss_hosts.c:/* vim: set expandtab ts=4 sw=4: */
```

Please add the new `ldap_iphost_*` options to `src/config/SSSDConfig/sssdoptions.py`, `src/config/cfg_rules.ini`, `src/config/etc/sssd.api.d/sssd-{ad|ipa|ldap}.conf` to make the config checker and the python config API aware of them.

I tested the patches in different setup and the tests worked well.

bye,
Sumit

scabrero commented:
Hi @sumit-bose 

I have fixed the issues, but I have skipped adding the `ldap_iphost_*` options to `src/config/etc/sssd.api.d/sssd-ipa.conf` because the resolver provider is not implemented yet for the IPA backend. I don't know it very well and after having a quick look to `src/providers/ipa/ipa_common.c` I am not sure if the schema is compatible with RFC2307 iphosts and there are design decisions to make like the CN to place the hosts. I will be happy to implement it if required, but I would like to get this merged and implement the netwoks lookup before.

sumit-bose commented:
Hi,

I think the following is missing to make the tests pass:

```
diff --git a/src/config/SSSDConfigTest.py b/src/config/SSSDConfigTest.py
index 74b4c8e..5e6de8e 100755
--- a/src/config/SSSDConfigTest.py
+++ b/src/config/SSSDConfigTest.py
@@ -845,9 +845,10 @@ class SSSDConfigTestSSSDDomain(unittest.TestCase):
             'ipa': ['id', 'auth', 'access', 'chpass', 'sudo', 'autofs',
                     'session', 'hostid', 'subdomains'],
             'ad': ['id', 'auth', 'access', 'chpass', 'sudo', 'autofs',
-                   'subdomains'],
+                   'subdomains', 'resolver'],
             'local': ['id', 'auth', 'chpass'],
-            'ldap': ['id', 'auth', 'access', 'chpass', 'sudo', 'autofs'],
+            'ldap': ['id', 'auth', 'access', 'chpass', 'sudo', 'autofs',
+                     'resolver'],
             'krb5': ['auth', 'access', 'chpass'],
             'proxy': ['id', 'auth', 'chpass'],
             'simple': ['access'],
@@ -1414,6 +1415,7 @@ class SSSDConfigTestSSSDConfig(unittest.TestCase):
             'chpass_provider',
             'sudo_provider',
             'subdomains_provider',
+            'resolver_provider',
             'default_shell',
             'fallback_homedir',
             'cache_credentials',

```

About IPA, I think a dedicated resolver is not needed.

bye,
Sumit

scabrero commented:
Thanks, I should have noticed... All tests passing now.

sumit-bose commented:
Thanks, sorry for nitpicking, but `ldap_iphost_entry_usn` is still missing in some files:

```
$ git grep -E '(ldap_iphost_name|ldap_iphost_entry_usn)' ../
../src/config/SSSDConfig/sssdoptions.py:        'ldap_iphost_name': _('IP host name attribute'),
../src/config/cfg_rules.ini:option = ldap_iphost_name
../src/config/etc/sssd.api.d/sssd-ad.conf:ldap_iphost_name = str, None, false
../src/config/etc/sssd.api.d/sssd-ldap.conf:ldap_iphost_name = str, None, false
../src/man/sssd-ldap-attributes.5.xml:                    <term>ldap_iphost_name (string)</term>
../src/providers/ad/ad_opts.c:    { "ldap_iphost_name", "cn", SYSDB_NAME, NULL },
../src/providers/ad/ad_opts.c:    { "ldap_iphost_entry_usn", NULL, SYSDB_USN, NULL },
../src/providers/ldap/ldap_opts.c:    { "ldap_iphost_name", "cn", SYSDB_NAME, NULL },
../src/providers/ldap/ldap_opts.c:    { "ldap_iphost_entry_usn", NULL, SYSDB_USN, NULL },
```

bye,
Sumit

scabrero commented:
Fixed, I have skipped `sssd-ldap-attributes.5.xml` because other `ldap_*_entry_usn` are not there either. I have also fixed a couple of issues I found while working in  the networks support.

sumit-bose commented:
Hi,

> Fixed, I have skipped `sssd-ldap-attributes.5.xml` because other `ldap_*_entry_usn` are not there either.

That's ok, thanks.

> I have also fixed a couple of issues I found while working in the networks support.

I guess there is a fair chance that there still might be some copy-and-paste issues like the ones you fixed.

Nevertheless, ACK.

Thank you for you patience.

bye,
Sumit

scabrero commented:
Sumit,

the networks support is almost finished and I caught the (I think) last errors, pushed again.

Thank you for you patience too.

pbrezina commented:
I removed accepted label since another review needs to be done.

sumit-bose commented:
Hi,

thanks for the update, I'm fine with the changes. ACK.

bye,
Sumit

pbrezina commented:
* `master`
    * 8a66d6e5a8abc9a89e12a5ef386541193f965b19 - TESTS: Add LDAP resolver target integration tests
    * e980b0f6a0aa2ded35f7cc81d8049eac159319da - PROXY: Add support for iphost enumeration
    * 45dbaddde02ac0c5b1af612ece19933d50c132a4 - AD: Implement iphost cleanup for expired cache entries
    * ae6d042cbb6b8a2bfecc37edeb3a1509796f6846 - LDAP: Implement iphost cleanup for expired cache entries
    * 10d9346afd1a415cb6c624e9134c977bc6aec8f8 - AD: Add support for iphost enumeration
    * 2be80a00cbd2227e828adb510c2d11cd3cfede48 - AD: Setup resolver enumeration tasks
    * 82b808d93ea8ffaea7a89f9196ad87a6f67441a9 - LDAP: Add support for iphost enumeration
    * 8a51bc0df6d9f45d00b561dc3f03373d0527ce96 - LDAP: Setup resolver enumeration tasks
    * 8b96109ffff823766ee5795f76bf17907392f6ee - CACHE_REQ: Add support for enumerating hosts
    * 11cc32e48b5cb70aecbe040b6629482cb497108e - SYSDB: Add support for enumerating hosts
    * bbcd849a4208bb54d83ebdb4b57e35f25b887dd7 - NSS: Add client support for [set|get|end]hostent()
    * 29b27395f39cb64834a054d1477467c045feaee3 - LDAP: Add support for iphost lookups (no enumeration)
    * b8fba016681aabf82eb7739259e323242aa4a8c4 - LDAP: Prepare for iphost lookups
    * 0498591eaeea0ff92e12e26e818fbf660d45c5b2 - AD: Initialize ldap_iphost_* options
    * 6a7775263cec46b577d3d25be2e2086136c7a134 - LDAP: Document new ldap_iphost_* options
    * a61c6d61c39e566eeda38ceceb8fa11831822d51 - LDAP: Initialize ldap_iphost_* options
    * 1402f10046eb830c9fcb591c8898f73dd13ddc95 - AD: Initialize resolver provider
    * 5672d2beb735d35bbf33efc25c48de4f31c33563 - LDAP: Initialize resolver provider
    * 29c583b6450af52fd0b93b526fda67b946d925ba - PROXY: Handle resolver hosts by address requests
    * 00bc7897163a9fed2bc7ba613f36c6b076e6d8c0 - SYSDB: Extend sysdb_store_host() to accept extra attributes
    * bbb7a45dfd16a7eb8e1b56da508a536e520e3bcd - PROXY: Store results from NSS library call into the cache
    * be791978953d0cfec64a08d7da2aae6d44af8e53 - PROXY: Handle resolver hosts by name requests
    * b1fe85eb0034f3aacb549322b4620976f09e39af - PROXY: Register resolver hosts handler method
    * 688e6a6b56966200e530c9e3a6de534e1f763b7e - PROXY: Load resolver NSS library
    * 0ec8bd578b55564182a21b05773c109393c16840 - PROXY: Create a module context to store id and auth contexts
    * b523fb6a03dbf6289159d4fb8b491fe35f1c2693 - UTIL: Allow to specify mandatory and optional symbols when loading nss libs
    * d6d03aafc49615ae1e4e88a692aa893214c3926d - CONFDB: Add a new resolver_timeout to timeout cached resolver entries
    * 6f6900374520b34b25199259731e410be7b7ebf3 - CONFDB: Add new options for resolver provider
    * 2c317ce9fc5a66ef82969baaf1d949fbee5dcd75 - DP: Implement resolver target handler
    * 014cd3a542bbb27d72c160a5c596b6797c6ed46e - TESTS: Add gethostbyname and gethostbyaddr NSS responder tests
    * 55cfacfe33e656840aed8bdb3ca7870a7bdad4dc - NSS: Add gethostbyname and gethostbyaddr support to the NSS responder
    * e931f27dfd3800c3931c6aba7e75e7499b64a913 - NSS: Add client support for hosts (non-enumeration)
    * 6e66e321797e52d7a9ed093683c45c678c9deec5 - CACHE_REQ: Implement ip_host_by_addr and ip_host_by_name plugins
    * dafdd066ef2358ee06796ce5bba8f0b0891c9d64 - CACHE_REQ: Add a data field to store network addresses
    * 1cb209556adad75264802de543f03e08f7058d23 - CACHE_REQ: Rename cache req host by name name plugin used by SSH
    * 469891df69faeceba2084efd8ce421b1ec11103e - RESPONDER: Add sss_dp_resolver_get_send
    * d76d818cb2015c4d5a4b70fac055f99134c5c858 - DP: Add a new filter type, filter by address
    * 99ce117106b9c0d0e0167f1c10f5840a7912fa7f - SBUS: Add new resolver target interface
    * 622849279be17cc9b7b59f81753aa05a0594df47 - SYSDB: Add index for hostAddress attribute
    * 860c457063fedf4c0020e705060b97cd5a80d1a7 - SYSDB: Add sysdb functions for hosts entries
    * d263fa9d643197bf0f3245e06149f7a3dd96f08d - UTIL: Add a function to canonicalize IP addresses
    * 12bd3f96ca101192ddc6b0afb424c183e68989c9 - STAP: Add missing session data provider target

pbrezina commented:
Thank you for getting this done. Excellent work!
